### ISSUE TITLE
Include failing property in violation json

### ISSUE THREAD

**Issue by ben-manes** [everit-org/json-schema#272]
I am migrating from the older [json-schema-validator](https://github.com/java-json-tools/json-schema-validator). The detail of a violation is slightly less detailed. For example in yours, 
```json
{
  "schemaLocation": "#",
  "pointerToViolation": "#",
  "causingExceptions": [
    {
      "schemaLocation": "#/properties/document",
      "pointerToViolation": "#/document",
      "causingExceptions": [],
      "keyword": "required",
      "message": "required key [name] not found"
    }
  ],
  "keyword": "allOf",
  "message": "#: only 1 subschema matches out of 2"
}
```

whereas the older library outputs:
```json
{
  "level": "error",
  "schema": {
    "loadingURI": "/1.0/entities/metadata/document.json#",
    "pointer": "/properties/document"
  },
  "instance": {
    "pointer": "/document"
  },
  "domain": "validation",
  "keyword": "required",
  "message": "object has missing required properties ([\"name\"])",
  "required": [
    "name"
  ],
  "missing": [
    "name"
  ]
}
```

To extract the `name` property I will have to parse the `message`, rather than the key being available directly.

**Comment by ben-manes** [comment]
It is actually a little inconsistent. Going through the translation process I get the full path when using the wrong type.

```json
{
    "schemaLocation": "#/properties/document/properties/name",
    "pointerToViolation": "#/document/name",
    "causingExceptions": [],
    "keyword": "type",
    "message": "expected type: String, found: Integer"
}
```

**Comment by erosb** [comment]
> Going through the translation process I get the full path when using the wrong type.
Sorry I don't understand this part.

**Comment by ben-manes** [comment]
Sorry, I meant that sometimes the pointer reflects the actual field. The difference in this case is that `required` is specified at the `properties` level unlike type-specific metadata like the `type` (`required` was like this in draft3, but doing so was too rigid). So the pointer is correct but not detailed enough, which is why the older library provided more detail in its json results.

My ask is to make all useful information programmatically available in through exception without users having to parse the `message` text.

**Comment by erosb** [comment]
Hello @ben-manes ,

so there are 2 separate requests in this ticket:

> To extract the name property I will have to parse the message, rather than the key being available directly.

I will keep this in mind when the error reporting gets reworked next time. That is likely to happen after the next version of the specification gets released & published, since there is an upcoming standardized format of validation errors.

> So the pointer is correct but not detailed enough

The pointer is actually correct, because making a property `"required"` is a constraint / validation criterion to the object, and not the property. Ie. your schema means that the "document" object _should have_ a "name" property - in other words if the "name" property doesn't exist the the "document" object is invalid. At least this is what the currently implemented error says and I think this approach also conforms with the spec. Also, I suspect there are already many users of this library who parse the current json pointer (hence they rely on the current behavior), therefore changing the pointer would break their code. Taking this into account I'm afraid it would cause much more pain than gain to change the pointer.

**Comment by ben-manes** [comment]
Yes, I’m not asking for backwards incompatible changes but enough details to easily extract out all of the information. If this is planned then you can close this.

**Comment by bex1111** [comment]
Is there any news on these?

**Comment by doko123** [comment]
Still no changes. The problem is that current solution forces to check every field in schema to look. for the incorrect value instead of having it explicitly in error.
I believe this is more bug than the feature:
`"None is not of type 'integer'\n\nFailed validating 'type' in schema[0]:\n    {'type': 'integer'}\n\nOn instance:\n    None`
vs sth like:
`"None is not of type 'integer'\n\nFailed validating 'type' in 'id_object':\n    {'type': 'integer'}\n\nOn instance:\n    None"`

**Comment by erosb** [comment]
Hello @doko123 this error message doesn't look any familiar for me. Probably you are looking at the wrong project.

**Comment by doko123** [comment]
Sory for the mistake I thought it was diff project :D
